### PR TITLE
`UnitControl`: refactor unit tests to TypeScript and modern RTL

### PR DESCRIPTION
**Pull request by ciampo**
## What?

This PR refactors `UnitControl`'s unit tests from JS to TypeScript, and partially rewrites some tests to use modern techniques (mostly using `@testing-library`'s methods).

This PR is a WIP. TODO:
- fix loading custom matchers definitions in TypeScript
- fix package.json configuration (can't add dev dependencies to a package's `package.json`)
- bonus: investigate why some `input`'s have `role="spinbutton`' and some have `role"textbox"` 

## Why?

This is part of the `@wordpress/components`'s TypeScript migration effort (https://github.com/WordPress/gutenberg/issues/35744).

Using TypeScript unit tests (instead of JS) is a great way to type-check our components from the point of view of their consumers, and thus catch early bugs and API inconsistencies.

## How?

- TBD

## Testing Instructions

- Project builds correctly
- Tests pass
- `UnitControl` works as expected in Storybook and in the block editor

### CODE REVIEW

File: packages/components/src/unit-control/test/index.tsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /**
  * External dependencies
  */
+import { describe, expect, it } from '@jest/globals';
```
Comment (ciampo) [inline packages/components/src/unit-control/test/index.tsx]: When working in JS, these globals are already in scope. In TypeScript, I had to explicitly include them (which, honestly, I don't mind since it feels less _magic_)

File: packages/components/package.json [hunk 2]
```diff
@@ -69,6 +69,10 @@
 		"reakit": "^1.3.8",
 		"uuid": "^8.3.0"
 	},
+	"devDependencies": {
+		"@jest/globals": "27.5.1",
+		"@testing-library/react": "11.2.2"
+	},
```
Comment (ciampo) [inline packages/components/package.json]: It looks like I can't list `devDependencies` in a package's `package.json`, but at the same time I have to list them in this file if I want to import them explicitly (e.g. importing `@jest/globals` in the unit test file).

What's the best solution here? Moving them to the root package.json in the `devDependencies` and then listing them again in this file's `dependencies`?

Comment (ciampo) [inline packages/components/package.json:75]: It looks like I can't list `devDependencies` in a package's `package.json`, but at the same time I have to list them in this file if I want to import them explicitly (e.g. importing `@jest/globals` in the unit test file).

What's the best solution here? Moving them to the root package.json in the `devDependencies` and then listing them again in this file's `dependencies`?

Comment (sirreal) [reply packages/components/package.json]: I don't have a definitive answer as I'm not familiar with how deps are set up now.

I can say we shouldn't list these as direct `dependencies` of this package, that would cause consumers to download them when they're absolutely not necessary to _use_ this package.

**Comment by sarayourfriend** [comment]
Based on testing-library's own instructions (https://github.com/testing-library/jest-dom#with-typescript) we need to be able to include the `test/unit/config/testing-library.js` into our tsconfig.

This presents some problems for the current monorepo tsconfig situation as we rely on `rootDir` usually to be some equivalent of `.`.

It also means everything we import in there needs to be TS enabled (mostly this relates to the `to-match-style-diff-snapshot` module).

You'll need to refactor all the existing tsconfig's to move the rootDir to the root of the repository otherwise TypeScript will complain that you're including a file in the configuration that is outside the rootDir for each individual tsconfig. You _could_ only do this for the components package and include the `test/unit/config/testing-library.ts` there, but it wouldn't solve the overall infrastructure issue for the project.

I also tried seeing if it was possible to add the `@testing-library/jest-dom` types into the `types` array, essentially making the types globally available for all projects using the base tsconfig (this is a hack and is discouraged by current TypeScript configuration best-practices) but that didn't work either.

I think the best route is the following (to be done in a separate PR probably):

1. Convert `to-match-style-diff-snapshot` to TypeScript
2. Convert `test/unit/config/testing-library.js` to TypeScript
3. Add `test/unit/config/testing-library.ts` to the `tsconfig.base.json` `include` array and update the `exclude` referencing test namespaces to read `**/test/**/*.js` so that it only excludes `js` files and doesn't squash our inclusion of the `testing-library.ts` file.
4. Update all the `tsconfig.json` for all packages to have a rootDir of the repository root (e.g., either `../..` or `..`) and update any paths that need updating within those tsconfig.json files.

After that I think the matches would start working. In my brief experiment the types were at least getting included and some of the errors went away. I'll note that there are also some legitimate errors, missing `jest` import and some implicit any's.

File: packages/components/src/unit-control/test/index.tsx (hunk 1 above)
Comment (sirreal) [reply packages/components/src/unit-control/test/index.tsx]: The globals are still in scope, TypeScript just doesn't know about them. We need to add `jest` to the `types` array so that its types are loaded. We never import `jest` so its types are never loaded (unless we put them in the `types` array:

https://github.com/WordPress/gutenberg/blob/9b0e9e411df3db10772ddb67d44d380bd3007934/packages/components/tsconfig.json#L6

But this `tsconfig` file currently represents this entire project, it would make these globals become global which isn't accurate.

The way I've structured this before is to make 2 projects (2 tsconfigs):
- `tsconfig.json`: the application code, excludes tests files (something like `**/tests/` in this case?)
- `tsconfig.tests.json` - only includes tests files. This is where we include the jest and testing-library types. The test project includes a reference to (depends on) the main project.

**Comment by sirreal** [comment]
I pushed a fix for the issues where TypeScript didn't have global types.

Ideally, I think tests would be treated as a separate project to avoid polluting the main project with globals that don't exists, but it's not a big risk. I described the setup a bit here https://github.com/WordPress/gutenberg/pull/39341#discussion_r824663947

**Comment by ciampo** [comment]
Thank you @sarayourfriend and @sirreal for the help so far!

After a couple of @sirreal 's commits, it looks like we've got a direction: creating separate `tsconfig` for tests file, which doesn't `emit` and that includes all of the additional types for `jest` / `testing-library` / etc..

For some reasons, I had to add `@emotion/core` to the newly created `tsconfig`, to prevent a `property 'css' is missing in type 'UnitControl'...` error (although I don't really understand why it's needed).

At the moment, running `npx tsc -p tsconfig.test.json` in the repo's root doesn't output any errors 🤞 

Although my VSCode still shows a TS error for the `toMatchDiffSnapshot` — for some reason,[ requiring `snapshot-diff` ](https://github.com/WordPress/gutenberg/pull/39341/commits/21a7e5202833ad1f5f25801ff5a81fac424a7d5a)in the testing library's setup fixes it 🤷  So, for some reason, it seems like VSCode is not able to load the new test TSConfig (not 100% related, but I found some interesting conversations going on https://github.com/jaredpalmer/tsdx/issues/871#issuecomment-695822562).

Since IDE support is important, I wonder if we should rework how we expose/extend our configs (I found [this comment](https://github.com/jaredpalmer/tsdx/issues/638#issuecomment-605728767)), or that we try the approach described by @sarayourfriend ?

To recap, I think I managed to push a couple of commits that fix issues, but they seem hacky and I don't really know why they worked 😅 

Any thoughts?

**Comment by sirreal** [comment]
I've reviewed the tsconfig changes and it seems like a reasonable setup. I think this is a good starting point 👍 

One nice thing is that the setup for non-test code is practically untouched. That's been working pretty well and these changes shouldn't disrupt it.

Certainly, there will be improvements to this setup but this seems like the place to start.

One side note: I noticed the TypeScript version used in this project is a couple versions old. It would be nice to get that dependency updated.

**Comment by ciampo** [comment]
Closing this PR in favour of #39436, where I polished the approach taken in this PR and removed the changes to `UnitControl`'s unit tests.

> One side note: I noticed the TypeScript version used in this project is a couple versions old. It would be nice to get that dependency updated.

I'll see if I can work on that in the next days.
